### PR TITLE
Improve the way to clean up storage devices for sandbox

### PR DESCRIPTION
**Pull request by jiangliu**
Currently there's no way to return error when cleaning up storage devices for sandbox. So add a way to return error and refine the logic to clean up resources.

### CODE REVIEW

**Review by lifupan — APPROVED** [review]

File: src/agent/src/storage/mod.rs [hunk 1]
```diff
@@ -55,6 +54,86 @@ pub struct StorageContext<'a> {
...
+impl StorageDevice for StorageDeviceGeneric {
+    fn path(&self) -> &str {
+        &self.path
+    }
+
+    fn cleanup(&self) -> Result<()> {
+        if !self.initialized {
+            return Ok(());
+        } else if self.path.is_empty() {
+            return Err(anyhow!("mountpoint path is empty"));
```
Comment (quanweiZhou) [inline src/agent/src/storage/mod.rs:99]: If the path is empty, there is no need to clean it up, maybe return `ok()` is better.

File: src/agent/src/storage/mod.rs [hunk 2]
```diff
@@ -55,6 +54,86 @@ pub struct StorageContext<'a> {
...
+/// An implementation of generic storage device.
+pub struct StorageDeviceGeneric {
+    path: String,
+    initialized: bool,
+}
+
+impl std::fmt::Debug for StorageDeviceGeneric {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("StorageDeviceGeneric")
+            .field("path", &self.path)
```
Comment (quanweiZhou) [inline src/agent/src/storage/mod.rs:66]: Lack of the field `initialized`.

File: src/agent/src/storage/mod.rs [hunk 3]
```diff
@@ -55,6 +54,86 @@ pub struct StorageContext<'a> {
     sandbox: &'a Arc<Mutex<Sandbox>>,
 }
 
+/// An implementation of generic storage device.
+pub struct StorageDeviceGeneric {
+    path: String,
+    initialized: bool,
```
Comment (quanweiZhou) [inline src/agent/src/storage/mod.rs:60]: Can you give more detail for the purpose of the field of `initialized`, currently only `StorageDeviceGeneric::default()` sets the `initialized = false`, What circumstances do `StorageDeviceGeneric::default()` and `StorageDeviceGeneric::new("")` has a difference.

File: src/agent/src/storage/mod.rs [hunk 4]
```diff
@@ -55,6 +54,86 @@ pub struct StorageContext<'a> {
...
+        }
+        if matches!(is_mounted(&self.path), Ok(true)) {
+            return Err(anyhow!("failed to umount mountpoint {}", self.path));
+        }
+
+        if path.is_dir() {
+            let is_empty = path.read_dir()?.next().is_none();
+            if !is_empty {
+                return Err(anyhow!("directory is not empty when clean up storage"));
+            }
```
Comment (quanweiZhou) [inline src/agent/src/storage/mod.rs:119]: Maybe the 116-119 is unnecessary because the `remove_dir` can ensure that the directory is empty.

**Review by quanweiZhou — COMMENTED** [review]
Thanks, @jiangliu, few comments.

File: src/agent/src/storage/mod.rs [hunk 5]
```diff
@@ -55,6 +54,86 @@ pub struct StorageContext<'a> {
...
+            // "remove_dir" will fail if the mount point is backed by a read-only filesystem.
+            // This is the case with the device mapper snapshotter, where we mount the block device
+            // directly at the underlying sandbox path which was provided from the base RO kataShared
+            // path from the host.
+            let _ = fs::remove_dir(path);
+        } else if !path.is_file() {
+            // TODO: should we remove the file for bind mount?
```
Comment (quanweiZhou) [inline src/agent/src/storage/mod.rs:126]: If the future such as the use of oss, maybe exist bind mount for files.

File: src/agent/src/storage/mod.rs (hunk 3 above)
Comment (jiangliu) [reply src/agent/src/storage/mod.rs]: I have refactor the code to use `path: Option<String>` to encode three states:
- uninitialized: None
- empty path: Some("")
- non-empty path: Some(path)

File: src/agent/src/storage/mod.rs (hunk 1 above)
Comment (jiangliu) [reply src/agent/src/storage/mod.rs]: Fixed

File: src/agent/src/storage/mod.rs (hunk 2 above)
Comment (jiangliu) [reply src/agent/src/storage/mod.rs]: Fixed

File: src/agent/src/storage/mod.rs (hunk 4 above)
Comment (jiangliu) [reply src/agent/src/storage/mod.rs:103]: There may be multiple filesystems mounted at the same destination directory, and umount() only unmounts one level, so hope to give a clear error message for such a case.

File: src/agent/src/storage/mod.rs (hunk 5 above)
Comment (jiangliu) [reply src/agent/src/storage/mod.rs:110]: Need to work to ensure that "deleting a data file" is safe:)

**Review by quanweiZhou — APPROVED** [review]
Thanks, @jiangliu, LGTM.

**Pull request merged**